### PR TITLE
FEDX-60 : Updated Null safety

### PR DESCRIPTION
**Pull request by tusharmathur-wk**
Migrate to null safety. Run the null safety migration tool with Dart migrate and check/adjust its suggestions if needed.
Apply the migration, and fix any static errors. Do a dart format, run the tests, fixing any issues.

### CODE REVIEW

File: lib/src/http/auto_retry.dart [hunk 1]
```diff
@@ -180,9 +180,10 @@ class RequestAutoRetry extends AutoRetryConfig {
 }
 
 @visibleForTesting
-Duration getRetryTimeoutThreshold(Duration timeoutThreshold, int numAttempts) {
+Duration? getRetryTimeoutThreshold(
+    Duration? timeoutThreshold, int numAttempts) {
   if (numAttempts <= 0) return timeoutThreshold;
-  var threshold = timeoutThreshold * numAttempts;
+  var threshold = timeoutThreshold! * numAttempts;
```
Comment (taylorweil-wk) [inline lib/src/http/auto_retry.dart:186]: this needs a null check or be made non-nullable

File: lib/src/http/browser/request_mixin.dart [hunk 2]
```diff
@@ -45,34 +45,29 @@ abstract class BrowserRequestMixin implements BaseRequest, CommonRequest {
...
-      // - content-length
-      final headersToAdd = Map<String, String>.from(finalizedRequest.headers);
-      headersToAdd.remove('connection');
-      headersToAdd.remove('content-length');
-      headersToAdd.forEach(_request.setRequestHeader);
-    }
+    final headersToAdd = Map<String, String>.from(finalizedRequest.headers);
+    headersToAdd.remove('connection');
+    headersToAdd.remove('content-length');
+    headersToAdd.forEach(_request!.setRequestHeader);
```
Comment (taylorweil-wk) [inline lib/src/http/browser/request_mixin.dart:51]: should we be doing a null check here? if not maybe we want to make this late non-nullable

File: lib/src/http/common/multipart_request.dart [hunk 3]
```diff
@@ -76,7 +76,7 @@ abstract class CommonMultipartRequest extends CommonRequest
     return '$senderPrefix${String.fromCharCodes(boundaryChars)}';
   }
 
-  String get boundary {
+  String? get boundary {
```
Comment (taylorweil-wk) [inline lib/src/http/common/multipart_request.dart:79]: can this be non-nullable?

**Review by taylorweil-wk — COMMENTED** [review]
got some last suggestions and questions, once resolved im good to approve. Sorry for the delay

File: lib/src/mocks/mock_http.dart [hunk 4]
```diff
@@ -278,21 +280,21 @@ class MockHttpInternal {
...
       if (handlersByMethod.containsKey(method)) {
-        handler = handlersByMethod[method];
+        handler = handlersByMethod[method!];
       } else if (handlersByMethod.containsKey('*')) {
         handler = handlersByMethod['*'];
       }
     }
     if (handler == null) return null;
-    return _RequestHandlerMatch(handler,
+    return _RequestHandlerMatch(handler as Function,
```
Comment (evanweible-wf) [inline lib/src/mocks/mock_http.dart:293]: We can avoid this cast if we update the `Object` types above to `Function`

File: lib/src/web_socket/global_web_socket_monitor.dart [hunk 5]
```diff
@@ -28,10 +26,10 @@ GlobalWebSocketMonitor newGlobalWebSocketMonitor() =>
     GlobalWebSocketMonitor._();
 
 WebSocketConnectEvent newWebSocketConnectEvent(
-        {@required String url,
-        @required bool wasSuccessful,
-        String sockJsSelectedProtocol,
-        List<String> sockJsProtocolsWhitelist}) =>
+        {required String url,
+        required bool? wasSuccessful,
```
Comment (evanweible-wf) [inline lib/src/web_socket/global_web_socket_monitor.dart:30]: I think we can make this non-nullable if we update the calling code to use `late`

File: _test_server/lib/src/handlers/example/http/cross_origin_credentials_handlers.dart [hunk 6]
```diff
@@ -27,8 +27,8 @@ Map<String, Handler> exampleHttpCrossOriginCredentialsRoutes = {
   '$pathPrefix/credentialed': CredentialedRequestHandler()
 };
 
-String session;
-String generateSessionCookie() {
+String? session;
+String? generateSessionCookie() {
```
[the remark this answers is not in the document]
Comment (evanweible-wf) [reply _test_server/lib/src/handlers/example/http/cross_origin_credentials_handlers.dart:31]: Whoops, never mind - it's possible `isValidSession()` would be called first, in which case we'd be reading an uninitialized late var. We can still make this function return non-nullable, but the `session` var will need to stay nullable.

File: _test_server/lib/src/handlers/test/http/custom.dart [hunk 7]
```diff
@@ -26,7 +26,7 @@ class CustomHandler extends Handler {
   @override
   Future<Null> get(HttpRequest request) async {
     request.response.statusCode =
-        int.parse(request.uri.queryParameters['status']) ?? HttpStatus.ok;
+        int.parse(request.uri.queryParameters['status']!);
```
[the remark this answers is not in the document]
Comment (evanweible-wf) [reply _test_server/lib/src/handlers/test/http/custom.dart:29]: I think I see now - `int.parse` now returns a non-nullable type, but we can't guarantee that the `status` query param exists. We should do this instead:
```suggestion
    final queryStatusCode = request.uri.queryParameters['status'];
    request.response.statusCode =
        queryStatusCode != null ? int.parse(queryStatusCode) : HttpStatus.ok;
```

**Review by evanweible-wf — CHANGES_REQUESTED** [review]
Thanks for doing this - this was a big migration! I'm going to put up a PR with my suggested changes, it took a bit of trial and error to get to the point where the request dispatch methods could return a non-null response.

File: example/pubspec.yaml [hunk 8]
```diff
@@ -4,7 +4,7 @@ description: Transport library for sending HTTP requests and opening WebSockets.
 homepage: https://github.com/Workiva/w_transport
 
 environment:
-  sdk: ">=2.11.0 <3.0.0"
+  sdk: ">=2.12.0 <3.0.0"
```
Comment (evanweible-wf) [inline example/pubspec.yaml:7]: This subpackage isn't actually ready to be migrated because it depends on over_react. Will have to revert this change.

File: test/unit/http/request_test.dart [hunk 9]
```diff
@@ -217,10 +218,10 @@ void _runCommonRequestSuiteFor(String name,
...
     });
 
-    test('DELETE (streamed)', () async {
-      MockTransports.http.expect('DELETE', requestUri);
-      await requestFactory().streamDelete(uri: requestUri);
-    });
+    // test('DELETE (streamed)', () async {
+    //   MockTransports.http.expect('DELETE', requestUri);
+    //   await requestFactory().streamDelete(uri: requestUri);
+    // });
```
Comment (evanweible-wf) [inline test/unit/http/request_test.dart:224]: Commented out test

File: test/unit/http/request_test.dart [hunk 10]
```diff
@@ -290,20 +291,6 @@ void _runCommonRequestSuiteFor(String name,
...
-      request.abort();
-      Future future = request.get(uri: requestUri);
-      expect(future, throwsA(isA<transport.RequestException>()));
-      await future.catchError((_) {});
-      expect(request.isDone, isTrue,
-          reason: 'canceled request should be marked as "done"');
-      expect(request.done, completes,
-          reason:
-              'canceled request should trigger completion of `done` future');
-    });
```
Comment (evanweible-wf) [inline test/unit/http/request_test.dart:305]: Why was this test deleted?

File: test/unit/http/request_test.dart [hunk 11]
```diff
@@ -312,7 +299,12 @@ void _runCommonRequestSuiteFor(String name,
       await Future.delayed(Duration(milliseconds: 100));
       request.abort();
       expect(future, throwsA(isA<transport.RequestException>()));
-      await future.catchError((_) {});
+      try {
+        await future;
+        fail('Expected RequestException, but future completed successfully.');
+      } catch (error) {
+        expect(error, isA<transport.RequestException>());
+      }
```
Comment (evanweible-wf) [inline test/unit/http/request_test.dart:307]: This works, but seems to duplicate the assertion just above where we are also asserting that it throws. Do we have to do this instead of using `.catchError` because of non-nullability?

**Review by evanweible-wf — CHANGES_REQUESTED** [review]

**Review by taylorweil-wk — APPROVED** [review]
love to see it 🥇

**Review by evanweible-wf — APPROVED** [review]

**Comment by robbecker-wf** [comment]
QA+1 CI passes and consumer testing looks good

**Comment by robbecker-wf** [comment]
@Workiva/release-management-p

**Comment by evanweible-wf** [comment]
+1 security

**Comment by robbecker-wf** [comment]
security +1 No logic changes around withCredentials

**Comment by robbecker-wf** [comment]
semver +1 we're doing a major

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
